File: frontend/src/app/core/services/entrenamiento.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import {
  Entrenamiento, EntrenamientoRequest,
  EntrenamientoEjercicio, EntrenamientoEjercicioRequest,
  Sesion, SesionRequest,
  EntrenamientoEspecifico, AsignarClienteRequest,
  TipoEntrenamiento
} from '../models/entrenamiento.model';
import { environment } from '../../../environments/environment';

@Injectable({ providedIn: 'root' })
export class EntrenamientoService {

  private base = environment.apiUrl;

  constructor(private http: HttpClient) {}

  // ─── Entrenamientos ───────────────────────────────────────────────────────

  listarPorGimnasio(gimnasioId: number, tipo?: TipoEntrenamiento): Observable<Entrenamiento[]> {
    let params = new HttpParams();
    if (tipo) params = params.set('tipo', tipo);
    return this.http.get<Entrenamiento[]>(`${this.base}/gimnasios/${gimnasioId}/entrenamientos`, { params });
  }

  obtener(id: number): Observable<Entrenamiento> {
    return this.http.get<Entrenamiento>(`${this.base}/entrenamientos/${id}`);
  }

  crear(gimnasioId: number, request: EntrenamientoRequest): Observable<Entrenamiento> {
    return this.http.post<Entrenamiento>(`${this.base}/gimnasios/${gimnasioId}/entrenamientos`, request);
  }

  actualizar(id: number, request: EntrenamientoRequest): Observable<Entrenamiento> {
    return this.http.put<Entrenamiento>(`${this.base}/entrenamientos/${id}`, request);
  }

  eliminar(id: number): Observable<void> {
    return this.http.delete<void>(`${this.base}/entrenamientos/${id}`);
  }

  // ─── Ejercicios del entrenamiento ─────────────────────────────────────────

  agregarEjercicio(entrenamientoId: number, request: EntrenamientoEjercicioRequest): Observable<EntrenamientoEjercicio> {
    return this.http.post<EntrenamientoEjercicio>(`${this.base}/entrenamientos/${entrenamientoId}/ejercicios`, request);
  }

  actualizarEjercicio(entrenamientoId: number, eeId: number, request: EntrenamientoEjercicioRequest): Observable<EntrenamientoEjercicio> {
    return this.http.put<EntrenamientoEjercicio>(`${this.base}/entrenamientos/${entrenamientoId}/ejercicios/${eeId}`, request);
  }

  eliminarEjercicio(entrenamientoId: number, eeId: number): Observable<void> {
    return this.http.delete<void>(`${this.base}/entrenamientos/${entrenamientoId}/ejercicios/${eeId}`);
  }

  // ─── Sesiones ─────────────────────────────────────────────────────────────

  listarSesionesGimnasio(gimnasioId: number): Observable<Sesion[]> {
    return this.http.get<Sesion[]>(`${this.base}/gimnasios/${gimnasioId}/sesiones`);
  }

  listarSesionesPorEntrenamiento(entrenamientoId: number): Observable<Sesion[]> {
    return this.http.get<Sesion[]>(`${this.base}/entrenamientos/${entrenamientoId}/sesiones`);
  }

  crearSesion(entrenamientoId: number, request: SesionRequest): Observable<Sesion> {
    return this.http.post<Sesion>(`${this.base}/entrenamientos/${entrenamientoId}/sesiones`, request);
  }

  actualizarSesion(sesionId: number, request: SesionRequest): Observable<Sesion> {
    return this.http.put<Sesion>(`${this.base}/sesiones/${sesionId}`, request);
  }

  eliminarSesion(sesionId: number): Observable<void> {
    return this.http.delete<void>(`${this.base}/sesiones/${sesionId}`);
  }

  asignarClienteASesion(sesionId: number, request: AsignarClienteRequest): Observable<Sesion> {
    return this.http.post<Sesion>(`${this.base}/sesiones/${sesionId}/asignar-cliente`, request);
  }

  // ─── Inscripción cliente (GRUPAL) ─────────────────────────────────────────

  inscribirse(sesionId: number): Observable<void> {
    return this.http.post<void>(`${this.base}/sesiones/${sesionId}/inscripcion`, {});
  }

  cancelarInscripcion(sesionId: number): Observable<void> {
    return this.http.delete<void>(`${this.base}/sesiones/${sesionId}/inscripcion`);
  }

  // ─── Específicos ──────────────────────────────────────────────────────────

  asignarEspecifico(entrenamientoId: number, request: AsignarClienteRequest): Observable<EntrenamientoEspecifico> {
    return this.http.post<EntrenamientoEspecifico>(`${this.base}/entrenamientos/${entrenamientoId}/asignar-cliente`, request);
  }

  listarSesionesCliente(clienteId: number): Observable<Sesion[]> {
    return this.http.get<Sesion[]>(`${this.base}/clientes/${clienteId}/sesiones`);
  }

  listarEspecificosCliente(clienteId: number): Observable<EntrenamientoEspecifico[]> {
    return this.http.get<EntrenamientoEspecifico[]>(`${this.base}/clientes/${clienteId}/entrenamientos-especificos`);
  }

  marcarCompletado(especificoId: number, completado: boolean): Observable<EntrenamientoEspecifico> {
    return this.http.patch<EntrenamientoEspecifico>(
      `${this.base}/entrenamientos-especificos/${especificoId}/completado?completado=${completado}`, {}
    );
  }
}

File: backend/src/main/java/com/fuerzapp/service/EntrenamientoService.java
package com.fuerzapp.service;

import com.fuerzapp.dto.AsignarClienteRequest;
import com.fuerzapp.dto.EntrenamientoEjercicioRequest;
import com.fuerzapp.dto.EntrenamientoEjercicioResponse;
import com.fuerzapp.dto.EntrenamientoEspecificoResponse;
import com.fuerzapp.dto.EntrenamientoRequest;
import com.fuerzapp.dto.EntrenamientoResponse;
import com.fuerzapp.dto.SesionRequest;
import com.fuerzapp.dto.SesionResponse;
import com.fuerzapp.enums.TipoEntrenamiento;

import java.util.List;

public interface EntrenamientoService {

    List<EntrenamientoResponse> listarPorGimnasio(Long gimnasioId, TipoEntrenamiento tipo);
    List<EntrenamientoResponse> listarPorGimnasioYEntrenador(Long gimnasioId, Long entrenadorId);
    void reasignarEntrenador(Long entrenamientoId, Long nuevoEntrenadorId);
    EntrenamientoResponse obtenerPorId(Long id);
    EntrenamientoResponse crear(Long gimnasioId, EntrenamientoRequest request);
    EntrenamientoResponse actualizar(Long id, EntrenamientoRequest request);
    void eliminar(Long id);

    List<EntrenamientoEjercicioResponse> listarEjercicios(Long entrenamientoId);
    EntrenamientoEjercicioResponse agregarEjercicio(Long entrenamientoId, EntrenamientoEjercicioRequest request);
    EntrenamientoEjercicioResponse actualizarEjercicio(Long entrenamientoId, Long eeId, EntrenamientoEjercicioRequest request);
    void eliminarEjercicio(Long eeId);

    List<SesionResponse> listarSesiones(Long entrenamientoId);
    List<SesionResponse> listarSesionesFuturasPorGimnasio(Long gimnasioId);
    SesionResponse crearSesion(Long entrenamientoId, SesionRequest request);
    SesionResponse actualizarSesion(Long sesionId, SesionRequest request);
    void eliminarSesion(Long sesionId);

    void inscribirseEnSesion(Long sesionId);
    void cancelarInscripcion(Long sesionId);
    SesionResponse asignarClienteASesion(Long sesionId, AsignarClienteRequest request);

    EntrenamientoEspecificoResponse asignarEspecifico(Long entrenamientoId, AsignarClienteRequest request);
    EntrenamientoEspecificoResponse marcarCompletado(Long especificoId, boolean completado);

    List<SesionResponse> listarSesionesPorCliente(Long clienteId);
    List<EntrenamientoEspecificoResponse> listarEspecificosPorCliente(Long clienteId);
}
